File: solve_cave.py
from collections import deque

def parse_input(filename):
    with open(filename) as f:
        lines = f.read().splitlines()
    water_units = int(lines[0].strip())
    cave_lines = [line for line in lines[2:] if line.strip()]
    grid = [list(line) for line in cave_lines]
    return grid, water_units

def fill_cave(grid, water_units):
    height, width = len(grid), len(grid[0])
    inlet_col = next((i for i, c in enumerate(grid[0]) if c == '~' or c == ' '), 1)
    # Place initial water unit at inlet
    r = 0
    while r + 1 < height and grid[r + 1][inlet_col] == ' ':
        r += 1
    grid[r][inlet_col] = '~'
    water_units -= 1
    def next_to_water(i, j):
        return j > 0 and grid[i][j - 1] == '~'
    def underneath_water(i, j):
        return i > 0 and grid[i - 1][j] == '~'
    for _ in range(water_units):
        placed = False
        for i in range(height - 1, -1, -1):
            for j in range(width - 1, -1, -1):
                if grid[i][j] == ' ':
                    if underneath_water(i, j) or next_to_water(i, j):
                        grid[i][j] = '~'
                        placed = True
                        break
            if placed:
                break
    return grid

def measure_depths(grid):
    height, width = len(grid), len(grid[0])
    depths = []
    for c in range(width):
        depth = 0
        flowing = False
        for r in range(height):
            if grid[r][c] == '~':
                depth += 1
                if r+1 < height and grid[r+1][c] == ' ':
                    flowing = True
                    break
            elif grid[r][c] == '#':
                break
        depths.append('~' if flowing else depth)
    return depths

def main():
    import sys
    input_file = sys.argv[1] if len(sys.argv) > 1 else 'simple_cave.txt'
    grid, water_units = parse_input(input_file)
    filled = fill_cave(grid, water_units)
    depths = measure_depths(filled)
    print(' '.join(str(d) for d in depths))

if __name__ == '__main__':
    main()

File: src/typescript/solve_cave.ts
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

interface ParsedInput {
  grid: string[][];
  waterUnits: number;
}

function parseInput(filename: string): ParsedInput {
  const content = readFileSync(filename, 'utf-8');
  const lines = content.split('\n');
  const waterUnits = parseInt(lines[0]?.trim() || '0');
  const caveLines = lines.slice(2).filter(line => line.length > 0);
  const grid = caveLines.map(line => line.split(''));
  return { grid, waterUnits };
}

function fillCave(grid: string[][], waterUnits: number): string[][] {
  // The input file already has one water unit placed, so we add (waterUnits - 1) more
  for (let i = 0; i < waterUnits - 1; i++) {
    addWater(grid);
  }
  return grid;
}

function addWater(grid: string[][]): void {
  const height = grid.length;
  const width = grid[0]?.length || 0;
  const queue: [number, number][] = [];
  
  // Find all current water cells
  for (let i = 0; i < height; i++) {
    for (let j = 0; j < width; j++) {
      if (grid[i]?.[j] === '~') {
        queue.push([i, j]);
      }
    }
  }
  
  // If no water yet, start at inlet (top left open cell)
  if (queue.length === 0) {
    for (let j = 0; j < width; j++) {
      if (grid[0]?.[j] === ' ') {
        grid[0]![j] = '~';
        return;
      }
    }
    return;
  }
  
  // Try to expand water from each cell in queue (BFS order)
  while (queue.length > 0) {
    const [i, j] = queue.shift()!;
    
    // Flow down
    if (i + 1 < height && grid[i + 1]?.[j] === ' ') {
      grid[i + 1]![j] = '~';
      return;
    }
    
    // Spread right
    if (j + 1 < width && grid[i]?.[j + 1] === ' ') {
      grid[i]![j + 1] = '~';
      return;
    }
    
    // Spread left
    if (j - 1 >= 0 && grid[i]?.[j - 1] === ' ') {
      grid[i]![j - 1] = '~';
      return;
    }
    
    // Spread up (rare, for trapped air)
    if (i - 1 >= 0 && grid[i - 1]?.[j] === ' ') {
      grid[i - 1]![j] = '~';
      return;
    }
  }
}

function measureDepths(grid: string[][]): (number | string)[] {
  const height = grid.length;
  const width = grid[0]?.length || 0;
  const depths: (number | string)[] = [];
  
  for (let c = 0; c < width; c++) {
    // Check if water is flowing in this column (has '~' with space below)
    let flowing = false;
    for (let r = 0; r < height - 1; r++) {
      if (grid[r]?.[c] === '~' && grid[r + 1]?.[c] === ' ') {
        flowing = true;
        break;
      }
    }
    
    if (flowing) {
      depths.push('~');
    } else {
      // Count all '~' symbols in the column
      let depth = 0;
      for (let r = 0; r < height; r++) {
        if (grid[r]?.[c] === '~') {
          depth++;
        }
      }
      depths.push(depth);
    }
  }
  
  return depths;
}

function main() {
  const args = process.argv.slice(2);
  const __dirname = dirname(fileURLToPath(import.meta.url));
  const defaultInput = join(__dirname, '../../data/input/simple_cave.txt');
  const inputFile = args[0] || defaultInput;
  
  const { grid, waterUnits } = parseInput(inputFile);
  const filled = fillCave(grid, waterUnits);
  const depths = measureDepths(filled);
  
  console.log(depths.join(' '));
}

if (require.main === module) {
  main();
}

export { parseInput, fillCave, measureDepths };
